### PR TITLE
PR for security check

### PR DESCRIPTION
**Pull request by boubaker**

### CODE REVIEW

File: analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+  @Override
+  public void start() {
+    // Can't be job, because the mapping retrival must be executed on each
+    // cluster node
+    esMappingUpdater.scheduleAtFixedRate(() -> {
+      try {
+        retrieveMapping(true);
+      } catch (Throwable e) {
+        if (LOG.isDebugEnabled()) {
+          LOG.warn("Error while checking connection status to Etherreum Websocket endpoint", e);
```
Comment (vsellier) [inline analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: The level is not in relation with the if.
You shoud not swallow the exception in this case IMO

Comment (vsellier) [inline analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: ``Etherreum Websocket`` ?

File: analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+
+      storeFieldsMappings();
+    } catch (Exception e) {
+      LOG.error("Error getting mapping of analytics", e);
+    }
+    return new HashSet<>(esMappings.values());
+  }
+
+  @Override
+  public ChartDataList compueChartData(AnalyticsFilter filter) {
```
Comment (vsellier) [inline analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]:
```suggestion
  public ChartDataList computeChartData(AnalyticsFilter filter) {
```

File: analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+    }
+  }
+
+  @Override
+  public List<StatisticData> retrieveData(AnalyticsFilter searchFilter) {
+    StringBuilder esQuery = buildFilterQuery(searchFilter, false);
+
+    String esQueryString = esQuery.toString();
+    esQueryString = fixJSONStringFormat(esQueryString);
+    LOG.debug("ES query to compute search count with filter {}: \n{}", searchFilter, esQueryString);
```
Comment (vsellier) [inline analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: Please avoid logging in several lines as much as possible

File: analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+                                      boolean useSort) {
+    // If it's a count query, no need for results and no need for sort neither
+    if (isCount) {
+      esQuery.append("     \"size\" : 0");
+      if (useSort) {
+        esQuery.append(",");
+      }
+    } else {
+      if (offset > 0) {
+        esQuery.append("     \"from\" : ").append(offset).append(",\n");
```
Comment (vsellier) [inline analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: Are the ``\n``s really necessary ?
IMO we should not add any json formatting constraint on the code.

File: analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+      @SuppressWarnings("rawtypes")
+      Iterator keys = jsonObject.keys();
+      while (keys.hasNext()) {
+        String key = keys.next().toString();
+        String fieldMappingString = jsonObject.getString(key);
+        StatisticFieldMapping fieldMapping = fromJsonString(fieldMappingString, StatisticFieldMapping.class);
+        esMappings.put(key, fieldMapping);
+      }
+    } catch (JSONException e) {
+      LOG.error("Error reading ES mapped fields");
```
Comment (vsellier) [inline analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: missing the exception

File: analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+  @Override
+  public void stop() {
+    queueProcessingExecutor.shutdown();
+  }
+
+  @Override
+  public void put(StatisticData data) {
+    StatisticDataQueueEntry statisticDataQueueEntry = new StatisticDataQueueEntry(data);
+    this.statisticDatas.put(statisticDataQueueEntry.getId(), statisticDataQueueEntry);
```
Comment (vsellier) [inline analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java]: You should add a protection to limit the number of records in memory.
**blocker for production**

File: analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import org.exoplatform.container.ExoContainerContext;
+import org.exoplatform.container.PortalContainer;
+import org.exoplatform.container.component.RequestLifeCycle;
+import org.exoplatform.services.log.ExoLogger;
+import org.exoplatform.services.log.Log;
+
+public class DummyStatisticDataQueueService implements StatisticDataQueueService, Startable {
+  private static final Log                   LOG                     = ExoLogger.getLogger(DummyStatisticDataQueueService.class);
+
+  private Map<Long, StatisticDataQueueEntry> statisticDatas          = Collections.synchronizedMap(new HashMap<>());
```
Comment (vsellier) [inline analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java]: why using a map here? Wouldn't a Queue be more efficient?

File: analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    List<StatisticDataQueueEntry> queueEntries = new ArrayList<>(this.statisticDatas.values());
+    LOG.debug("Processing {} documents", queueEntries.size());
+    statisticDataProcessorService.process(queueEntries);
+    for (StatisticDataQueueEntry statisticDataQueueEntry : queueEntries) {
+      if (statisticDataQueueEntry.isProcessed()) {
+        this.statisticDatas.remove(statisticDataQueueEntry.getId());
+      }
+    }
+    if (this.statisticDatas.size() > 0) {
+      LOG.warn("Queue has {} remaining documents not processed successfully", this.statisticDatas.size());
```
Comment (vsellier) [inline analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java]: Won't this log be logged all the time as events will be added during the indexation ?

File: analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    }
```
Comment (vsellier) [inline analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java]: Is it possible to have execution time exposed in jmx ?

File: pom.xml [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+  <description>eXo Analytics Addon</description>
+  <modules>
+    <module>analytics-api</module>
+    <module>analytics-services</module>
+    <module>analytics-listeners</module>
+    <module>analytics-webapps</module>
+    <module>analytics-packaging</module>
+  </modules>
+  <scm>
+    <connection>scm:git:git://github.com/exoplatform/poc-sandbox.git</connection>
```
Comment (vsellier) [inline pom.xml]: Wrong url

File: pom.xml [hunk 11]
```diff
@@ -0,0 +1,117 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <artifactId>addons-parent-pom</artifactId>
+    <groupId>org.exoplatform.addons</groupId>
+    <version>9</version>
```
Comment (vsellier) [inline pom.xml]: Should be 10 at least

Comment (boubaker) [reply pom.xml]: done

File: pom.xml (hunk 10 above)
Comment (boubaker) [reply pom.xml]: done

File: analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java (hunk 9 above)
Comment (boubaker) [reply analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java]: done

File: analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java (hunk 8 above)
Comment (boubaker) [reply analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java]: deleted

File: analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java (hunk 7 above)
Comment (boubaker) [reply analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java]: converted to a local cache

File: analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java (hunk 6 above)
Comment (boubaker) [reply analytics-services/src/main/java/org/exoplatform/analytics/queue/service/DummyStatisticDataQueueService.java]: converted to a local cache

File: analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java (hunk 5 above)
Comment (boubaker) [reply analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: done

File: analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java (hunk 4 above)
Comment (boubaker) [reply analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: done

File: analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java (hunk 3 above)
Comment (boubaker) [reply analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: done

File: analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java (hunk 2 above)
Comment (boubaker) [reply analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: done

File: analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java (hunk 1 above)
Comment (boubaker) [reply analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: done

Comment (boubaker) [reply analytics-services/src/main/java/org/exoplatform/analytics/es/service/ESAnalyticsService.java]: done

File: analytics-api/pom.xml [hunk 12]
```diff
@@ -0,0 +1,68 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>org.exoplatform.addons.analytics</groupId>
+    <artifactId>analytics-parent</artifactId>
+    <version>1.0.x-SNAPSHOT</version>
+  </parent>
+  <artifactId>analytics-api</artifactId>
+  <name>eXo Add-on:: eXo Analytics - API</name>
```
[the remark this answers is not in the document]
Comment (boubaker) [reply analytics-api/pom.xml]: done

**Pull request merged**